// @ziggler/clanker/dist/tools/dynamic/pwd.js
import React from "react";
import { Text } from "ink";
import { createTool, ToolCategory } from "../../registry";
const pwdTool = createTool().id("pwd").name("Print Working Directory").description("Get the current working directory path").category(ToolCategory.FileSystem).tags("filesystem", "navigation", "directory", "pwd").examples([
  {
    description: "Get current working directory",
    arguments: {},
    result: "/Users/username/projects/my-app"
  }
]).renderResult(({ isExecuting, result }) => {
  if (isExecuting) {
    return /* @__PURE__ */ React.createElement(Text, { color: "cyan" }, "\u23BF Getting current directory...");
  }
  if (!result?.success) {
    return /* @__PURE__ */ React.createElement(Text, { color: "red" }, "\u23BF ", result?.error || "Failed to get working directory");
  }
  return /* @__PURE__ */ React.createElement(Text, { color: "gray" }, "\u23BF ", result.output);
}).execute(async (args, context) => {
  const cwd = process.cwd();
  return {
    success: true,
    output: cwd,
    data: { path: cwd }
  };
}).build();
var pwd_default = pwdTool;
export {
  pwd_default as default
};
